### PR TITLE
improve merge-pr script

### PR DESCRIPTION
**Pull request by gares**
The script now performs many more checks and reports errors in
a more intelligible way.

### CODE REVIEW

**Comment by gares** [comment]
Example of execution:

```
gares@gargamel:~/COQ/coq$ ~/merge-pr.sh 7023
info: title for PR 7023 is [ssreflect] Respect Opaque in FO unification 
info: PR 7023 targets branch coq:master 
info: you are merging in master 
info: remote for master is upstream 
info: fetching from upstream the PR 
Fetching upstream
From https://github.com/coq/coq
 * branch                  master     -> FETCH_HEAD
From https://github.com/coq/coq
 * branch                  refs/pull/7023/head -> FETCH_HEAD
info: commit for for PR 7023 is 50409f26d4c4f6f71af20cbaea234b8c0649dd26 
error: CI failure on ["ci/circleci: vst"] 
Continue anyway? [y/N] y
info: merging 
Merge made by the 'recursive' strategy.
 plugins/ssrmatching/ssrmatching.ml4 | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)
```

**Comment by gares** [comment]
screenshot from 2018-03-21 16-11-09

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: What is the imperfection exactly?

Comment (gares) [reply dev/tools/merge-pr.sh]: We check that something changes in that directory, not that a new overlay is present.

I think the check should be refined to also print the overlays, so that one has a list of links to click for the tasks that are recommended (two lines below)

**Comment by gares** [comment]
@maximedenes I did not propose an update for MERGING as we discussed (the suggestion contained in the last warning about overlays is missing from that document).

**Comment by gares** [comment]
This file lacks a code owner, I can be one of them btw

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [inline dev/tools/merge-pr.sh]: I don't understand what is going on but isn't there a typo here (repetition of `for`)

**Comment by maximedenes** [comment]
> This file lacks a code owner

I'm on it.

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [inline dev/tools/merge-pr.sh]: Cool! Could you do the same for milestones? I'm tired of having to set milestones after a PR has been merged.

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: Should use `error` function for redness.

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: It could be cleaning overlays or something like this. Not worth to mention the imperfection IMHO.

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [inline dev/tools/merge-pr.sh]: Please use "they" instead of "he".

**Review by Zimmi48 — COMMENTED** [review]
So cool!

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
I think it would be cleaner to quote messages, eg `info title for PR $PR is ${BLUE}$TITLE` -> `info "title for PR $PR is ${BLUE}$TITLE"`.
Then the various message function should only use 1 argument, eg
```bash
info() {
  echo -e "${GREEN}info:${BLACK}" "$1" ${BLACK}
}
```

**Comment by Zimmi48** [comment]
> I did not propose an update for MERGING as we discussed (the suggestion contained in the last warning about overlays is missing from that document).

Might still be worth doing it in this PR.

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: Oh and also check that the PR has at least one "kind: " label :)

Comment (SkySkimmer) [reply dev/tools/merge-pr.sh]: >Oh and also check that the PR has at least one "kind: " label :)

Is that really something we want to check?
I don't understand what you want to happen with milestones as well.

**Comment by maximedenes** [comment]
Btw, a reminder in case someone merges a PR that touches this script: it is tricky, since merge overwrites the script while it is running...

File: dev/tools/merge-pr.sh
Comment (gares) [reply dev/tools/merge-pr.sh]: typo

**Comment by maximedenes** [comment]
> This file lacks a code owner, I can be one of them btw

Fixed by #7031.

**Comment by gares** [comment]
@SkySkimmer I did quote things
@Zimmi48 I did add a check for milestones, I hope this is what you had in mind

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: Missing quotes for `$@` and `$BLACK`

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: Missing quotes for $1

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: What I've been enforcing manually since the 8.7 release (or even since the 8.6.1 release) is that all merged PRs are in a milestone and have a kind. This means that one can browse all the changes that happened between two versions by checking the corresponding milestone (and the kind labels show what kind of change a PR represents). This is like a mega CHANGES file but with much more information.

Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: A further improvement that would be really great would be for the script to allow setting the milestone directly (without going back in the web interface). This should be doable but only with write access, which would require people to first get a personal API token, etc. Much more complex thus!

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: Do not use $@ in a string, instead use $*.

Comment (gares) [reply dev/tools/merge-pr.sh]: why shouldn't I?
```
gares@gargamel:~/COQ/coq$ cat a.sh 
foo(){
  echo "foo $@ bar"
}
foo bla bla "bla bla"
gares@gargamel:~/COQ/coq$ bash a.sh 
foo bla bla bla bla bar
```

File: dev/tools/merge-pr.sh
Comment (gares) [reply dev/tools/merge-pr.sh]: I'll add a sanity check for kind: too

**Comment by gares** [comment]
Now we have a missing "kind:stuff" check

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [reply dev/tools/merge-pr.sh]: Because it makes shellcheck complain https://github.com/koalaman/shellcheck/wiki/SC2145

Comment (gares) [reply dev/tools/merge-pr.sh]: No, this is the white space linter syndrome again.

Tell me what is wrong, not that a linter complains.
If there is a reason I'm fine making a fix. According to the doc of bash, my code is just fine.

Comment (JasonGross) [reply dev/tools/merge-pr.sh]: Alternatively, because if someone ever decides to get rid of the `${GREEN}info:${BLACK} ` prefix, and leaves `$@`, then passing something like `-n foo` to this function will result in `echo` printing without a newline:
```bash
jgross@jgross-Leopard-WS:~/Documents/repos/fiat-crypto$ cat a.sh
foo(){
    echo "$@ bar"
}
foo -n bla bla "bla bla"
jgross@jgross-Leopard-WS:~/Documents/repos/fiat-crypto$ bash a.sh
bla bla bla bla barjgross@jgross-Leopard-WS:~/Documents/repos/fiat-crypto$
```

**Comment by gares** [comment]
And now I've removed the perfectly working both according to my tests and the doc of bash `".. $@ .."`.

Long live the linters

**Review by SkySkimmer — APPROVED** [review]
I'll try to be a bit less `-Wpedantic` in the future.

File: dev/tools/merge-pr.sh
Comment (maximedenes) [inline dev/tools/merge-pr.sh]: Don't you want to `2 > &1`? Have you tested with a missing util?

File: dev/tools/merge-pr.sh
Comment (maximedenes) [inline dev/tools/merge-pr.sh]: Don't you want `^[1-9][0-9]*` instead?

File: dev/tools/merge-pr.sh
Comment (maximedenes) [inline dev/tools/merge-pr.sh]: Don't you want to print something for pending jobs?

Comment (maximedenes) [reply dev/tools/merge-pr.sh]: What about canceled jobs?

**Review by maximedenes — CHANGES_REQUESTED** [review]

File: dev/tools/merge-pr.sh
Comment (SkySkimmer) [reply dev/tools/merge-pr.sh]: >Don't you want to `2 > &1`? Have you tested with a missing util?

After you said that I looked up which specs (they don't exist, which isn't posix) and it seems macos which exits with 0 even for missing utilities https://stackoverflow.com/questions/762631/find-out-if-a-command-exists-on-posix-system
TLDR use `if ! command -v "$1" > /dev/null`

Comment (maximedenes) [reply dev/tools/merge-pr.sh]: Does it mean that the following line is also broken? https://github.com/coq/coq/blob/ef0fc7730df3cb585b621955010115407d80d2e7/Makefile.build#L211

Comment (gares) [reply dev/tools/merge-pr.sh]: posix epic fail:
```
$ gpg -v
gpg: WARNING: no command supplied.  Trying to guess what you mean ...
gpg: Go ahead and type your message ...
^C
gpg: signal Interrupt caught ... exiting
```

Comment (gares) [reply dev/tools/merge-pr.sh]: maybe we should try with:
```
echo masochistic UI | $command -v
```

Can you dismiss this change request?

Comment (SkySkimmer) [reply dev/tools/merge-pr.sh]: No I mean the builtin `command`

ie 
```sh
$ command -v bash
/bin/bash (or wherever your bash is)
$ echo $?
0
$ command -v foobarbaz
$ echo $?
1
```

**Comment by SkySkimmer** [comment]
Seems so.

Gaëtan Gilbert

File: dev/tools/merge-pr.sh
Comment (herbelin) [reply dev/tools/merge-pr.sh]: FWIW, on my OS X,  both buit-in csh `which` and `/usr/bin/which` returns 1 if the command does not exist.

File: dev/tools/merge-pr.sh
Comment (vbgl) [inline dev/tools/merge-pr.sh]: This line is flawed.

When I tried, I got two errors. First, I was on a branch with no corresponding remote. Then I moved to a branch with remote `origin` (aka my fork on github) and not `upstream` (aka the official repo on github).

Comment (gares) [reply dev/tools/merge-pr.sh]: I think I had the same problem,but I forgot to fix the script ;-)

Thanks (ask for a change please, so that I don't forget again)

File: dev/tools/merge-pr.sh
Comment (gares) [reply dev/tools/merge-pr.sh]: thanks I'll fix that

**Comment by gares** [comment]
should be all good now

**Comment by Zimmi48** [comment]
Thanks for the good work @gares.

**Comment by gares** [comment]
Thanks but please test test test...

**Comment by vbgl** [comment]
Why is `curl` used with option `-s`? I got errors, but could not guess which before I tryed without this `--silent` option…

**Comment by vbgl** [comment]
```
error: remote upstream does not point to the official Coq repo
error: it points to https://github.com/coq/coq.git instead
```

Although the error seems legit, the message seems wrong. Please add an hint that the SSH address should be used instead of the HTTPS one (if this is what is required).

**Comment by gares** [comment]
I fixed the error message, thanks for testing!

For `curl`, I don't know. We may `>>` to a log file, but I've the feeling we are almost ready to land on Mars with this script. I hope the tweaks to land on Europe can wait until 2020 (and be make in other PRs)

Maybe I should start to make some pressure on the assignee.... ;-)

**Comment by gares** [comment]
I wanted to write 2010, but it would have not made sense, so my brain subverted reality... if you like sf movies, I beg your pardon.

File: dev/tools/merge-pr.sh
Comment (maximedenes) [reply dev/tools/merge-pr.sh]: I think we can leave with it being mentioned.

**Review by maximedenes — APPROVED** [review]

**Comment by Zimmi48** [comment]
I am testing this merge script and I have a problem with the black color which isn't very visible on my dark screen. I'll send a screenshot.

**Comment by maximedenes** [comment]
@Zimmi48 I fixed it, will push.

**Comment by Zimmi48** [comment]
I also have this error message:

```
error: remote upstream does not point to the official Coq repo 
error: that is git@github.com:coq/coq.git 
error: it points to git@github.com:coq/coq instead 
```

which is useless because the only difference is the `.git` which actually doesn't matter.

**Comment by SkySkimmer** [comment]
What is the goal of the remote check?
If it's because we need the remote to get the PR commit maybe we could check the remote only if fetching the PR fails to say something like "fetching PR xx failed, your remote is YY but the official remote is ZZ".

**Comment by Zimmi48** [comment]
OK @maximedenes so no need for a screenshot then.

Other than that, this is great work! Thanks a lot @gares!

Finally, @maximedenes mentioned that it was complicated to merge PRs that changed the merge script because it changes a file while it is being executed. A way to avoid this problem is to checkout the branch in a different worktree and merge in master using the new script from the worktree. E.g. `../coq-pr-7029/dev/tools/merge-pr.sh XXXX`

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [inline dev/tools/merge-pr.sh]: This info predates this PR but it is incomplete.

**Comment by Zimmi48** [comment]
> If it's because we need the remote to get the PR commit maybe we could check the remote only if fetching the PR fails to say something like "fetching PR xx failed, your remote is YY but the official remote is ZZ".

It looks like this is the only use for the remote indeed.

File: dev/tools/merge-pr.sh
Comment (gares) [reply dev/tools/merge-pr.sh]: ok

File: dev/tools/merge-pr.sh
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: You meant "live" I guess.

Comment (maximedenes) [reply dev/tools/merge-pr.sh]: Yes, I'm thinking a bit too much about leaving for the week-end probably :)

**Comment by gares** [comment]
I think I have enough. I'm happy to review more PRs on the same script, but I'm not going to change this PR.

**Comment by gares** [comment]
I'm wondering about a "needs: merging" label

**Comment by Zimmi48** [comment]
The only blocking problem for me is this black. If you want to keep the black, I need you to change the background color as well, so that it is readable.

**Comment by gares** [comment]
It is fixed:
```
BLACK="\033[39m" 
```

**Comment by maximedenes** [comment]
Right, I'll fix it.

**Comment by maximedenes** [comment]
Thanks for the nice work, @gares !

**Pull request merged**
